cmd/collector/app/builder_flags.go

```diff
@@ -40,6 +40,7 @@ const (
 	collectorGRPCMaxReceiveMessageLength = "collector.grpc-server.max-message-size"
 	collectorMaxConnectionAge            = "collector.grpc-server.max-connection-age"
 	collectorMaxConnectionAgeGrace       = "collector.grpc-server.max-connection-age-grace"
+	collectorTenantHeader                = "collector.tenant-header"
```

Comment:
gRPC is not the only way to submit data to collector. Is the intention to require the same header name across transports?

Reply:
Do we need to specify tenants at the deployment time? I wanted to avoid restarting Jaeger (and changing deployments) when a new tenant is added.

Reply:
The configuration can be extended to accept tenants in other forms, like watching a file or by integrating with another source. This would be a good opportunity to start introducing a control plane where tenants can be provided externally via api call, such that the integration with the source can be decoupled from Jaeger code.

Reply:
I would propose tackling the tenant restrictions in a separate PR.

More sophisticated deployments should do tenant validation in the auth layer (proxy). Which makes me think do we even need Jaeger to do this validation? A valid requirement for predefined tenants might be if the storage implementation needs to do some initialization per tenant. In the case of ES it should not be required, but maybe other storages like C* need to create a new DB that cannot be done in-flight of the first store call.

Reply:
The validation I'm most concerned about is not whether tenant is valid, but whether it is sent at all. Without that we open a hole that could allow other tenants to read your data.

I agree that there needs to be a stronger authentication somewhere other than a header, which is trivial to spoof.

@jkowall any experience to share here?

Reply:
I will relax the handling of valid tenants so that if `multi_tenancy.tenants` is undefined any tenant will pass.  This will allow the spanwriter to reject spans.  In the future we can provide a way for the writer to reject earlier.

Storage might have rules about what a valid tenant can be that we can't express on a list.  For example, some storage might reject tenants that include non-ASCII characters.  Or some storage might want to hot-enable/disable a tenant.